### PR TITLE
Add template for Tensor Parallel (DTensor) + Ray Train

### PR DESCRIPTION
**Pull request by tohtana**
This PR adds a template that demonstrates how to use Ray Train with PyTorch's native DTensor for tensor parallelism.

At a high level, this template covers:
- Understanding 2D parallelism (Tensor Parallelism + Data Parallelism)
- Setting up a 2D device mesh with PyTorch
- Applying DTensor tensor parallelism to transformer layers
- Combining DTensor with FSDP2 for data parallelism
- TP-aware data loading to ensure correct gradient computation
- Saving and loading distributed checkpoints with Ray Train

Tested environment: 4 GPUs with 2-way TP × 2-way DP using Qwen2.5-0.5B.

### CODE REVIEW

File: doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.ipynb [hunk 1]
```diff
@@ -0,0 +1,713 @@
...
+    "\n",
+    "    if world_rank == 0:\n",
+    "        logger.info(f\"Device mesh created: {device_mesh}\")\n",
+    "\n",
+    "    # [2] Create model with random initialization on the target device\n",
+    "    dtype = get_mixed_precision_dtype()\n",
+    "    prev_device = torch.get_default_device()\n",
+    "    try:\n",
+    "        torch.set_default_device(device)\n",
+    "        model = AutoModelForCausalLM.from_config(hf_config).to(dtype=dtype)\n",
```
[the remark this answers is not in the document]
Comment (tohtana) [reply doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.ipynb]: adressed

File: doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.ipynb [hunk 2]
```diff
@@ -0,0 +1,717 @@
+{
+  "cells": [
+    {
+      "cell_type": "markdown",
+      "metadata": {},
+      "source": [
+        "# Get started with Tensor Parallelism (DTensor) and Ray Train\n",
```
Comment (JasonLi1909) [inline doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.ipynb]:
```suggestion
        "# Get started with PyTorch Tensor Parallelism and Ray Train\n",
```
Clarifies framework

File: doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.ipynb [hunk 3]
```diff
@@ -0,0 +1,717 @@
...
+        "\n",
+        "**Time to complete:** 20 min\n",
+        "\n",
+        "This template shows how to train large language models using tensor parallelism with PyTorch's native DTensor API and Ray Train for distributed execution.\n",
```
Comment (JasonLi1909) [inline doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.ipynb]:
```suggestion
        "This template shows how to train large language models using tensor parallelism with PyTorch's native [Distributed Tensor API](https://docs.pytorch.org/docs/stable/distributed.tensor.html) and Ray Train for distributed execution.\n",
```
Change to "Distributed Tensor" because we are using both `torch.distributed.tensor` (DTensor) and `torch.distributed.tensor.parallel` (parallelize_module, ColwiseParallel, etc.) and add link for reference

File: doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.ipynb [hunk 4]
```diff
@@ -0,0 +1,717 @@
...
+        "\n",
+        "**Tensor Parallelism (TP)** shards model weights across multiple GPUs, enabling training of models that are too large to fit on a single GPU. Combined with **Data Parallelism (DP)**, this creates a powerful **2D parallelism** strategy that scales efficiently to many GPUs.\n",
+        "\n",
+        "This tutorial provides a step-by-step guide covering:\n",
+        "\n",
+        "- Understanding 2D parallelism (Tensor Parallelism + Data Parallelism)\n",
+        "- Setting up a 2D device mesh with PyTorch\n",
```
Comment (JasonLi1909) [inline doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.ipynb]:
```suggestion
        "- Setting up a 2D PyTorch `DeviceMesh`\n",
```

**Review by JasonLi1909 — COMMENTED** [review]
Thanks! Super cool that we're showing how we can compose FSDP2 and tensor parallel. A few things:

- Maybe we should change the title to include FSDP2 because we're going beyond basic tensor parallel. Also, Ray Train provides the `prepare_model` API which wraps the model in the `DistributedDataParallel` . Provide context on why we use FSDP2 (DTensor) for data parallelism instead of the older and simpler `DistributedDataParallel` that has no model sharding.
- Briefly elaborate on how tensor parallel composes with FSDP, including which parameters live on which workers and how the two work together during forward, backward, and step.
- Explain when FSDP alone is not enough and when tensor parallel should be used (e.g. large individual layers are compute bounded). Not necessary, but as a bonus it might be cool to visualize the gains of using tensor parallel (e.g. GPU memory usage before and after).

File: doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.ipynb [hunk 5]
```diff
@@ -0,0 +1,717 @@
...
+        "</style>"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {},
+      "source": [
+        "## Understanding 2D Parallelism\n",
+        "\n",
+        "2D parallelism combines two complementary strategies:\n",
```
Comment (JasonLi1909) [inline doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.ipynb]: A small technicality, is 2D parallelism always referring to tp and dp or can it also be other pairs of parallelisms? If so, rephrase to: 

```suggestion
        "2D parallelism can be achieved by combining two complementary strategies:\n",
```

File: doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.ipynb (hunk 2 above)
Comment (tohtana) [reply doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.ipynb]: Considering your comments above, changed it to "Get started with 2D Parallelism (Tensor + Data Parallelism) using FSDP2 and Ray Train" to include "FSDP2". Does this look okay?

File: doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.ipynb (hunk 3 above)
Comment (tohtana) [reply doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.ipynb]: Avoid saying `DTensor` unless it is specifically about `DTensor` objects.

File: doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.ipynb (hunk 5 above)
Comment (tohtana) [reply doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.ipynb]: 2D parallelism doesn't always mean TP+DP. I clarified that we refer to TP+DP here by "2D Parallelism (Tensor Parallelism + Data Parallelism)"

File: doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.ipynb (hunk 4 above)
Comment (tohtana) [reply doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.ipynb]: Fixed as suggested

**Comment by tohtana** [comment]
Thank you for the comments! I replied to each suggestion in the code. For your remaining comments:

> Provide context on why we use FSDP2 (DTensor) for data parallelism instead of the older and simpler DistributedDataParallel that has no model sharding.

Explicitly explained why we need FSDP in "Understanding 2D Parallelism (Tensor Parallelism + Data Parallelism)"

> Explain when FSDP alone is not enough and when tensor parallel should be used (e.g. large individual layers are compute bounded). Not necessary, but as a bonus it might be cool to visualize the gains of using tensor parallel (e.g. GPU memory usage before and after).

I added a brief analysis about overhead of TP and FSDP in "When to use Tensor Parallelism vs FSDP".
Ideally we should show some numbers, but the performance significantly varies according to conditions (batch size, sequence length, model size, etc.) It seems a bit too much to show the complete benchmark varying those conditions. So I just added general ideas about the overheads and a common strategy.

> Briefly elaborate on how tensor parallel composes with FSDP, including which parameters live on which workers and how the two work together during forward, backward, and step.

Added a figure and explained the device placement of shards and communications for both methods.

File: doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.ipynb [hunk 6]
```diff
@@ -0,0 +1,742 @@
...
+        "\n",
+        "This tutorial provides a step-by-step guide covering:\n",
+        "\n",
+        "- Understanding 2D parallelism (Tensor Parallelism + Data Parallelism)\n",
+        "- Setting up a 2D PyTorch `DeviceMesh`\n",
+        "- Applying tensor parallelism with the Distributed Tensor API to transformer layers\n",
+        "- Combining with FSDP2 for data parallelism\n",
+        "- TP-aware data loading to ensure correct gradient computation\n",
+        "- Distributed checkpointing with Ray Train\n",
+        "\n",
```
Comment (liulehui) [inline doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.ipynb]: how about changing this to match the order of rest of the notebook?
- Understanding 2D parallelism (Tensor Parallelism + Data Parallelism) and 2D PyTorch DeviceMesh,
- Setting up torch dataloader to be compatible with 2D parallelism
- Prepare model with DTensor and FSDP2 API according to DeviceMesh
- Checkpointing
- Distributed Training with Ray Train

File: doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.ipynb [hunk 7]
```diff
@@ -0,0 +1,742 @@
...
+        "train_loop_config = {\n",
+        "    # Model and data\n",
+        "    \"model_name\": \"Qwen/Qwen2.5-0.5B\",  # Small model for demo\n",
+        "    \"dataset_name\": \"wikitext\",\n",
+        "    \"dataset_percentage\": 5.0,  # Use 5% of dataset for faster demo\n",
+        "    # Parallelism\n",
+        "    \"tp_size\": tp_size,\n",
+        "    \"dp_size\": dp_size,\n",
+        "    # Training hyperparameters\n",
+        "    \"batch_size\": 1,\n",
```
Comment (liulehui) [inline doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.ipynb]: just for my understanding, this batch_size is per dp_rank right? with dp_size = 2, essentially we are training with global_batch_size = 2 right?

**Review by liulehui — COMMENTED** [review]
❤️super useful content!!

File: doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.ipynb (hunk 6 above)
Comment (tohtana) [reply doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.ipynb]: Thank you, reordered the items

File: doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.ipynb (hunk 7 above)
Comment (tohtana) [reply doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.ipynb]: Great catch, that's correct. To clarify it, I renamed `batch_size` `batch_size_per_gpu` and added a paragraph to explain it.

**Review by liulehui — APPROVED** [review]

**Review by liulehui — APPROVED** [review]

File: doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.md [hunk 8]
```diff
@@ -0,0 +1,641 @@
...
+```bash
+%%bash
+pip install torch transformers datasets
+```
+
+
+```python
+# Enable Ray Train V2 for the latest train APIs
+import os
+os.environ["RAY_TRAIN_V2_ENABLED"] = "1"
```
Comment (matthewdeng) [inline doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.md]: Don't think we need to mention this anymore since it's default to True now.

File: doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.md [hunk 9]
```diff
@@ -0,0 +1,641 @@
+# Get started with 2D Parallelism (Tensor + Data Parallelism) using FSDP2 and Ray Train
+
+**Time to complete:** 20 min
+
+This template shows how to train large language models using tensor parallelism with PyTorch's native [Distributed Tensor API](https://docs.pytorch.org/docs/stable/distributed.tensor.html) and Ray Train for distributed execution.
```
Comment (matthewdeng) [inline doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.md]: Both `Distributed Tensor` and `DTensor` are used across the tutorial. For consistency I think you can introduce it once as `DTensor (Distributed Tensor)` and use `DTensor` shorthand throughout the tutorial, while reserving lowercase "distributed tensor" to describe the concept.

File: doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.md [hunk 10]
```diff
@@ -0,0 +1,641 @@
...
+print("Training completed successfully!")
+print(f"Final metrics: {result.metrics}")
+```
+
+## Scaling to larger models
+
+To train larger models like Qwen2-7B or Llama-3-8B, adjust the configuration:
+
+```python
+# For 8 GPUs: 4-way TP, 2-way DP
```
Comment (matthewdeng) [inline doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.md]: Move this code comment to be in the above text instead to show that this is just another example config, similar to what's done in the previous section.

File: doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.md [hunk 11]
```diff
@@ -0,0 +1,641 @@
...
+            logger.info("dp_size=1, skipping FSDP sharding (TP only)")
+
+    if world_rank == 0:
+        num_params = sum(p.numel() for p in model.parameters())
+        logger.info(f"Model initialized with {num_params:,} parameters")
+
+    return model, tp_mesh, dp_mesh, tp_rank, dp_rank
+```
+
+## 4. Checkpointing
```
Comment (matthewdeng) [inline doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.md:373]: Can we also show loading checkpoint?

**Review by matthewdeng — APPROVED** [review]
This is awesome.

File: doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.md (hunk 8 above)
Comment (tohtana) [reply doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.md]: Removed

File: doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.md (hunk 10 above)
Comment (tohtana) [reply doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.md]: Updated as suggested. Applied the same fix on AutoTP example.

Comment (tohtana) [reply doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.md]: Applied the same fix on AutoTP example.

File: doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.md (hunk 8 above)
Comment (tohtana) [reply doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.md]: Applied the same fix on AutoTP example.

File: doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.md (hunk 11 above)
Comment (tohtana) [reply doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.md:373]: Great catch! Added code for loading. Applied the same fix on AutoTP example.

File: doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.md (hunk 9 above)
Comment (tohtana) [reply doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.md]: Updated as suggested.

File: doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.ipynb [hunk 12]
```diff
@@ -0,0 +1,801 @@
...
+    "        model.load_state_dict(torch.load(model_path, map_location=device))\n",
+    "        optimizer.load_state_dict(torch.load(optimizer_path, map_location=device))\n",
+    "\n",
+    "        start_epoch = 0\n",
+    "        if os.path.exists(metadata_path):\n",
+    "            with open(metadata_path, \"r\") as f:\n",
+    "                metadata = json.load(f)\n",
+    "            start_epoch = metadata.get(\"epoch\", -1) + 1\n",
+    "\n",
+    "    return start_epoch\n"
```
[the remark this answers is not in the document]
Comment (tohtana) [reply doc/source/train/examples/pytorch/tensor_parallel_dtensor/README.ipynb:504]: Switched to DCP
